panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,732 @@
+
+  if (connection)
+    {
+      NMSettingUser *setting_user;
+      const gchar *value;
+
+      setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+      g_assert (setting_user != NULL);
+
+      value = nm_setting_user_get_data (setting_user, NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED);
+
+      gtk_switch_set_active (GTK_SWITCH (self->automatic_updates_switch), g_strcmp0 (value, "1") == 0);
+    }
+  else
+    {
+      gtk_switch_set_active (GTK_SWITCH (self->automatic_updates_switch), FALSE);
+    }
+
+  g_signal_handlers_unblock_by_func (self->automatic_updates_switch, on_automatic_updates_switch_changed_cb, self);
+}
+
+static void
+load_metered_label_from_connection (CcUpdatesPanel *self,
+                                    NMConnection   *connection)
+{
+  gtk_widget_set_visible (self->metered_data_label, connection != NULL);
+
+  if (connection)
+    {
+      NMSettingConnection *setting;
+      NMMetered metered;
+
+      setting = nm_connection_get_setting_connection (connection);
+      g_assert (setting != NULL);
+
+      metered = nm_setting_connection_get_metered (setting);
+
+      if (metered == NM_METERED_YES || metered == NM_METERED_GUESS_YES)
+        gtk_label_set_label (GTK_LABEL (self->metered_data_label), _("Limited data plan connection"));
+      else
+        gtk_label_set_label (GTK_LABEL (self->metered_data_label), _("Unlimited data plan connection"));
+    }
+}
+
+static void
+update_active_network (CcUpdatesPanel *self)
+{
+  NMAccessPoint *ap;
+  NMConnection *connection;
+  NMDevice *device;
+  const gchar *icon_name;
+
+  get_active_connection_and_device (self, &device, &connection, &ap);
+
+  /* Setup the new device */
+  if (self->current_device != device)
+    cleanup_signals (self);
+
+  if (g_set_object (&self->current_device, device) && device)
+    {
+      self->changed_id = g_signal_connect_swapped (device,
+                                                   "state-changed",
+                                                   G_CALLBACK (on_network_changed_cb),
+                                                   self);
+    }
+
+  /* Icon */
+  icon_name = get_network_status_icon_name (self, device, ap);
+  gtk_image_set_from_icon_name (GTK_IMAGE (self->network_status_icon), icon_name, GTK_ICON_SIZE_BUTTON);
+
+  /* Name */
+  if (ap)
+    {
+      GBytes *ssid = nm_access_point_get_ssid (ap);
+      g_autofree gchar *title = nm_utils_ssid_to_utf8 (g_bytes_get_data (ssid, NULL), g_bytes_get_size (ssid));
+
+      gtk_label_set_label (GTK_LABEL (self->network_name_label), title);
+    }
+  else
+    {
+      gtk_label_set_label (GTK_LABEL (self->network_name_label),
+                           connection ? _("Connected") : _("No active connection"));
+    }
+
+  /* This makes sure we never return NULL to 'nm_connection_get_setting (c, NM_TYPE_SETTING_USER)' */
+  ensure_setting_connection (self, connection);
+  ensure_setting_user (self, connection);
```

Comment:
I’m not 100% convinced these calls are sufficient to ensure you always have a `NMSettingConnection` or `NMSettingUser` when you need it. I think it would be more obviously correct to change `ensure_setting_connection()` to have the same declaration as `nm_connection_get_setting_user()`, replace all calls to the latter with the former, and have the implementation of `ensure_setting_connection()` be something like:

```
static NMSettingConnection *
ensure_setting_connection (NMConnection *connection)
{
  NMSettingConnection *setting = nm_connection_get_setting_connection (connection);
  if (setting == NULL)
    {
      setting = NM_SETTING_CONNECTION (nm_setting_connection_new ());
      nm_connection_add_setting (connection, NM_SETTING (setting));
    }

  return setting;
}
```

Same for `ensure_setting_user()`.